src/pstore/pstore.c

```diff
@@ -0,0 +1,399 @@
+                        }
+                }
+        }
+        if (dmesg) {
+                (void) write_dmesg(dmesg, dmesg_size, id);
+                free(dmesg);
+                dmesg = NULL;
+                dmesg_size = 0;
+        }
+}
+
+static int list_files(PStoreList *list, const char *sourcepath) {
+        DIR *dirp;
+        struct dirent *de;
+        PStoreEntry *pe;
+        int r = 0;
+
+        dirp = opendir(sourcepath);
+        if (!dirp)
+                return log_error_errno(errno, "Failed to opendir %s: %m", sourcepath);
+
+        FOREACH_DIRENT(de, dirp, return log_error_errno(errno, "Failed to iterate through %s: %m", sourcepath)) {
+                _cleanup_free_ char *ifd_path;
+
+                ifd_path = path_join(sourcepath, de->d_name);
+                if (!ifd_path)
+                        return log_oom();
+
+                _cleanup_free_ char *buf = NULL;
+                size_t buf_size;
+
+                /* Now read contents of pstore file */
+                r = read_full_file(ifd_path, &buf, &buf_size);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to read file %s: %m", ifd_path);
+                        continue;
+                }
+
+                if (!GREEDY_REALLOC(list->entries, list->n_entries_allocated, list->n_entries + 1))
+                        return log_oom();
+
+                pe = &list->entries[list->n_entries++];
+                pe->dirent = *de;
+                pe->content = TAKE_PTR(buf);
+                pe->content_size = buf_size;
+                pe->is_binary = true;
+                pe->handled = false;
+        }
+        closedir(dirp);
+
+        return r;
+}
+
+static int run(int argc, char *argv[]) {
+        _cleanup_(pstore_entries_reset) PStoreList list = {};
+        int r;
+
+        log_open();
+
+        /* Ignore all parse errors */
+        (void) parse_config();
+
+        log_debug("Selected storage '%s'.", pstore_storage_to_string(arg_storage));
+        log_debug("Selected Unlink '%d'.", arg_unlink);
+
+        if (arg_storage == PSTORE_STORAGE_NONE)
+                /* Do nothing, intentionally, leaving pstore untouched */
+                return 0;
+
+        /* Obtain list of files in pstore */
+        r = list_files(&list, arg_sourcedir);
+        if (r < 0)
+                return r;
+
+        /* Handle each pstore file */
+        if (list.n_entries > 0) {
+                /* Sort files lexigraphically ascending, generally needed by all */
+                qsort_safe(list.entries, list.n_entries, sizeof(PStoreEntry), compare_pstore_entries);
+
+                /* Process known file types */
+                process_dmesg_files(&list);
+
+                /* Move left over files out of pstore */
+                PStoreEntry *pe;
+                for (size_t n = 0; n < list.n_entries; n++) {
+                        pe = &list.entries[n];
+                        if (!pe->handled)
```

Comment:
i think it would be prettier to do this check inside of move_file(), i.e. as a heck early up that makes the function exit early...

Reply:
done